### ISSUE TITLE
Polynomials

### ISSUE THREAD

**Issue by avekens** [metamath/set.mm#1088]
On working with polynomials, I detected the following:
- The definition of eval is in section 10.10.1  "Definition and basic properties", but it is used in section 14.1.1  "Abstract polynomials, continued" for the first time! Furthermore, there is a section 10.10.2  "Polynomial evaluation" which does not contain any definition - how are the theorems of this section related to the definition of the evaluation?
- there are many definitions which are not used (in any theorem) yet:  evalSub (I am currently working on that, see also Google group https://groups.google.com/d/msg/metamath/4dHsn535ucg/RokgvX-VBgAJ), mHomP, mPSDer, selectVars, AlgInd - are these just placeholders? Shouldn't they be contained in a mathbox as long as they are not used?
- why are the theorems of section 14.1.1  "Abstract polynomials, continued" not contained in (a subsection of) section 10.10  "Abstract multivariate polynomials"? This section can be copied to the end of section 10.10 without causing any conflicts/problems...
- why is the definition of the degree of a polynomial based on the field of complex numbers (CCfld)? This seems to be the reason why it is not defined in section 10.10. If section 14.1.2  "Polynomial degrees" is moved up following 10.11.1 "definition and basic properties" (in which CCfld is defined), there are no broken dependencies anymore. The same holds for section 14.1.3 "The division algorithm for univariate polynomials".

Is there any reason for splitting up the definitions and theorems for abstract polynomials into sections contained in different parts? How should this be cleaned up?

**Comment by avekens** [comment]
Bullet points 1-3 shoule be resolved by PR #1093.
Bullet point 4 (definition of the degree of a polynomial based on the field of complex numbers (CCfld)) is still open...

**Comment by digama0** [comment]
I don't think that decoupling the definition of degree from CCfld is useful. If anything, it would be better to move CCfld earlier, maybe a lot earlier, and I forget what the barrier is but in any case it should be available for general use, because it makes arguments more uniform and allows us to apply algebraic reasoning on CC.

**Comment by avekens** [comment]
A(n intermediate) solution could be to swap sections 10.11 "The complex numbers as an algebraic extensible structure" and 10.10  "Abstract multivariate polynomials". I tried it out - there will be no conflicts/broken dependencies. Afterwards, sections 14.1.1 "Polynomial degrees" and 14.1.2  "The division algorithm for univariate polynomials" could be moved to the new section 10.11  "Abstract multivariate polynomials". Would this be OK?

Section "The complex numbers as an algebraic extensible structure" cannot be moved further (as a whole), because some of its definitions/theorems depend on definitions/theorems of section 10.9  "Associative algebras". I do not know if it makes sense to split up this section, extracting the definition of CCfld and related theorems (and additional, required theorems to prove them) into a seperate section which could be moved further.

**Comment by benjub** [comment]
This may not be completely on topic, but I think that globally, real and complex numbers are introduced too early in set.mm.  For a particular example: the binomial theorem for complex numbers ~binom should be a corollary of the binomial theorem ~srgbinom.  I think that concerning the ordering of the sections, Bourbaki's TOC is hard to beat.
One way I can see that forbids totally adhering  to it is for instance "coding" number systems so that the real numbers are genuinely included in the complex numbers, for instance.  But globally, put general algebraic stuff before the real and complex stuff.

**Comment by tirix** [comment]
But aren't the members of structures indexed by integer numbers, themselves complex numbers?
See ` Base = Slot 1 ` , ` +g = Slot 2 ` , etc.
Do you suggest to replace them with e.g. ordinal numbers?

**Comment by benjub** [comment]
That's what I was hinting at in my second paragraph: set.mm wants that natural numbers be genuinely included in complex numbers, and natural numbers are required early on (e.g. to define powers in groups), so there is some balancing to find between the mathematics and the "engineering". I do not have a ready made answer.  For structures, ordinal numbers might be a possibility, as you say... but maybe not worth it ?  You have more experience in formalizing proofs to tell.

**Comment by nmegill** [comment]
> This may not be completely on topic, but I think that globally, real and complex numbers are introduced too early in set.mm.

I think it depends on the audience we intend for set.mm.  Is it mainly for mathematicians, or is it to appeal more generally?

At least in my case as an engineer in grad school, analysis was the first "real" math course (i.e. one focusing on proofs).  We used Rudin's book, "Intro. to Math. Analysis" I think.  The first thing that was developed was the Dedekind construction of real and complex numbers, then some point-set topology, then applications to the usual trig, log, calculus.  In other words close to what we have done in set.mm.

> For a particular example: the binomial theorem for complex numbers ~binom should be a corollary of the binomial theorem ~srgbinom. 

For non-mathematicians, going through the layers building up from semirings to CCFld to get a proof of the binomial theorem is a rather severe requirement.  I would say most would much prefer to see a direct proof in CC.  If required to go through semirings many or most would end up abandoning the effort as far too abstract.

I'm not saying that we shouldn't use this approach for more advanced work, but I think basic theorems would have more appeal proved directly in CC.

> I think that concerning the ordering of the sections, Bourbaki's TOC is hard to beat.

While Bourbaki may have a certain elegance in its structure, I don't think it is often used as a textbook, and it may or may not be the best approach for set.mm.  (I found its set theory rather crazy. :)

**Comment by benjub** [comment]
> I think it depends on the audience we intend for set.mm. Is it mainly for mathematicians, or is it to appeal more generally?

I agree it's the main question, and I agree with the rest of @nmegill's post too. Bourbaki's Elements would be a terrible textbook ! And it was not intended by its authors to be a textbook, but rather a reference work (well, the initial project was a textbook, but it quickly became a project to write a reference work). I think that a repository of formalized proofs like set.mm is naturally more a reference work than a textbook, and I've always been skeptical about its pedagogical value to learn math.  But I agree this might be due to my background: I got acquainted to formalized mathematics, as a hobby, only after having a very mature understanding of mathematics and professional practice. And I have a hard time thinking the other way around is viable... but again, I wouldn't know for sure (it looks like it really helped some persons you tutored), so I suspend my judgment.

**Comment by benjub** [comment]
Agreed.

**Comment by avekens** [comment]
Main issues are solved, discussion if the degree of a polynomial based on the field of complex numbers (CCfld) is still open (decision can be postponed), general discussion about the structure of set.mm is a separate topic.
